File: backend/routes/authRoutes.js
// backend/routes/authRoutes.js
const express = require('express');
const router = express.Router();
const multer = require('multer');
const path = require('path');
const fs = require('fs');

// Import all 4 functions in ONE single block so there are no duplicates
const { 
    registerUser, 
    loginUser, 
    registerShopOwner, 
    loginShopOwner,
    getProfile,
    updateProfile,
    loginAdmin,
    updatePassword,
    registerShop
} = require('../controllers/authController');
const verifyToken = require('../middleware/authMiddleware');

const uploadDirectory = path.join(__dirname, '../uploads');
const storage = multer.diskStorage({
    destination: (req, file, cb) => cb(null, uploadDirectory),
    filename: (req, file, cb) => cb(null, Date.now() + '-' + file.originalname.replace(/\s+/g, '-'))
});
const upload = multer({ storage });

// --- User Routes ---
router.post('/register/user', registerUser);    
router.post('/login/user', loginUser);

// --- The secure profile route! ---
router.get('/me', verifyToken, getProfile);


// --- Shop Owner Routes ---
router.post('/register/shop', upload.single('logo'), registerShopOwner);
router.post('/login/shop', loginShopOwner);

// This will be used for both users and shop owners to update their profile info
router.put('/me', verifyToken, updateProfile); 

// --- Admin Route ---
router.post('/login/admin', loginAdmin);

// Update your routes to intercept the 'logo' file!
router.post('/register/shop', upload.single('logo'), registerShop); 
router.put('/profile', verifyToken, upload.single('logo'), updateProfile);

// --- NEW PROFILE ROUTES ---
router.get('/profile', verifyToken, getProfile);
router.put('/profile', verifyToken, updateProfile);
router.put('/password', verifyToken, updatePassword);

module.exports = router;

File: backend-python/routes/auth_routes.py
# backend-python/routes/auth_routes.py
from typing import Optional

from fastapi import APIRouter, Depends, File, Form, UploadFile

from controllers import auth_controller as ctrl
from middleware.auth import verify_token

router = APIRouter()


# --- User Routes ---
@router.post("/register/user", status_code=201)
async def register_user(
    name: str = Form(...),
    email: str = Form(...),
    password: str = Form(...),
    location: Optional[str] = Form(None),
):
    return await ctrl.register_user(name, email, password, location)


@router.post("/login/user")
async def login_user(
    email: str = Form(...),
    password: str = Form(...),
):
    return await ctrl.login_user(email, password)


@router.post("/login")
async def unified_login(
    email: str = Form(...),
    password: str = Form(...),
):
    return await ctrl.unified_login(email, password)


# --- Secure profile routes ---
@router.get("/me")
async def get_me(
    role: Optional[str] = None,
    current_user: dict = Depends(verify_token),
):
    return await ctrl.get_profile(current_user["id"], role)


@router.put("/me")
async def update_me(
    name: str = Form(...),
    location: str = Form(...),
    role: Optional[str] = Form(None),
    category_id: Optional[int] = Form(None),
    logo: Optional[UploadFile] = File(None),
    current_user: dict = Depends(verify_token),
):
    return await ctrl.update_profile(current_user["id"], name, location, role, category_id, logo)


# --- Shop Owner Routes ---
@router.post("/register/shop", status_code=201)
async def register_shop_owner(
    shop_name: str = Form(...),
    email: str = Form(...),
    password: str = Form(...),
    location: str = Form(...),
    category_id: Optional[int] = Form(None),
    logo: Optional[UploadFile] = File(None),
):
    return await ctrl.register_shop_owner(shop_name, email, password, location, category_id, logo)


@router.post("/login/shop")
async def login_shop_owner(
    email: str = Form(...),
    password: str = Form(...),
):
    return await ctrl.login_shop_owner(email, password)


# --- Admin Route ---
@router.post("/login/admin")
async def login_admin(
    email: str = Form(...),
    password: str = Form(...),
):
    return await ctrl.login_admin(email, password)


# --- Profile Routes ---
@router.get("/profile")
async def get_profile(
    role: Optional[str] = None,
    current_user: dict = Depends(verify_token),
):
    return await ctrl.get_profile(current_user["id"], role)


@router.put("/profile")
async def update_profile(
    name: str = Form(...),
    location: str = Form(...),
    role: Optional[str] = Form(None),
    category_id: Optional[int] = Form(None),
    logo: Optional[UploadFile] = File(None),
    current_user: dict = Depends(verify_token),
):
    return await ctrl.update_profile(current_user["id"], name, location, role, category_id, logo)


@router.put("/password")
async def update_password(
    currentPassword: str = Form(...),
    newPassword: str = Form(...),
    role: Optional[str] = Form(None),
    current_user: dict = Depends(verify_token),
):
    return await ctrl.update_password(current_user["id"], currentPassword, newPassword, role)
